### ISSUE TITLE
Validation NDCG is different from the one TFR model gives when called after training with the same ELWC

### ISSUE THREAD

**Issue by azagniotov** [tensorflow/ranking#201]
Hello Team,

I noticed that `NDCG@20` calculated using ranker predictions is different (i.e.: 20%-30% smaller) from to `NDCG@20` observed from the validation printout during training (printout provided below).

Some of the relevant parameters that I used for the training are as follows:
```
_LOSS_FUNCTION = tfr.losses.RankingLossKey.APPROX_NDCG_LOSS
_LIST_SIZE = 20
_GROUP_SIZE = 20
```

After the training, I get the following final validation printout:
```
INFO:tensorflow:Saving dict for global step 15000: global_step = 15000,
labels_mean = 0.10526316, logits_mean = -0.3222814, loss = -0.27790582,
metric/ndcg@10 = 0.28906482, metric/ndcg@20 = 0.28906482, metric/ndcg@5 = 0.0
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 15000: 
checkpoints_model_dir/model.ckpt-15000
INFO:tensorflow:Loss for final step: -0.71023804.

({'labels_mean': 0.10526316,
  'logits_mean': -0.3222814,
  'loss': -0.27790582,
  'metric/ndcg@10': 0.28906482,
  'metric/ndcg@20': 0.28906482,
  'metric/ndcg@5': 0.0,
  'global_step': 15000},
 [])
```

When the training was over, I called the model using the ranker object using the following code:
```
predictions = ranker.predict(input_fn=lambda: predict_input_fn(_VALIDATION_DATA_PATH))
```
`_VALIDATION_DATA_PATH` contains a path to TF record with a single `ELWC`.

In order to ensure consistent results in my investigation, I used the same ELWC stored as TF record under  `_VALIDATION_DATA_PATH` as an input to `eval_input_fn` .

That single `ELWC` contains twenty examples, where 11th example (index == 10) has a non-zero label value, e.g.:
```
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

The result of calling `ranker.predict` is shown below:
```
Predictions:  [-8.0555525  -3.6958659  -5.7905755  -1.6236236  -1.985581   -3.25897
 -2.3058856   0.24394287 -2.4218245  -0.24467875 -5.279163   -1.3242409
 -3.7202942  -4.9686694  -2.583102   -0.5809151  -7.3875494  -2.1003609
 -3.2946153  14.231506  ]
```

After sorting the above predictions in descending order and applying the same order to the labels, I get the following label sort:
```
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0]
```

After calculating the `NDCG@20` on the above labels, I got `0.239812466` as a result, which is much lower than `metric/ndcg@20': 0.28906482`, in the final validation printout. I used the same common formula for DCG as the one given in [metrics_impl.py](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/python/metrics_impl.py#L92).

Why would there be such a discrepancy?

Incidentally, I am able to obtain `ndcg@20: 0.28906482` when I calculate NDCG using the following label sort (please note the non-zero label on the `10th` place, index == 9) :
```
[0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

**Comment by xuanhuiwang** [comment]
Can you try _GROUP_SIZE = 1? This is the current recommended setting. When _GROUP_SIZE > 1, there are some randomness when shuffling the list.

**Comment by azagniotov** [comment]
Hi @xuanhuiwang,

Thank you for your reply. I tried with `_GROUP_SIZE = 1` and I can confirm that I consistently see the same value for NDCG@20 in every validation printout (i.e.: `metric/ndcg@20 = 0.25595802`) and after the training, I calculated the `NDCG@20` from the labels sorted by predictions to also be `0.25595802`.

My understanding was that if I want to benefit from the listwise scoring of the examples under the context, I should set the `_GROUP_SIZE > 2`. I also found Michael's clarification about various TFR sizes: https://github.com/tensorflow/ranking/issues/51

In what cases should the `_GROUP_SIZE` be set to more than `2`? Can you please clarify `_GROUP_SIZE = 1` to be the current recommended setting and why?

**Comment by azagniotov** [comment]
@xuanhuiwang 

Can you also please explain where exactly the randomness is when `_GROUP_SIZE > 1`? 

Is the example list randomly reshuffled every time the validation NDCG is computed? Or is it reshuffled randomly once during the training process? We have observed that the validation NDCG rises steadily and smoothly during the training process -- can we use it as an indication that our model is becoming better during training? 

If the validation NDCG is computed on a randomly reshuffled list, then it would seem like we could not interpret a steady improvement in the validation NDCG as an indication of our model getting better, which is a shame

**Comment by xuanhuiwang** [comment]
@azagniotov, when _GROUP_SIZE=1, each document is scored independently, you can benefit from using pairwise or listwise *losses*.

When _GROUP_SIZE > 1, more than 1 documents are scored together by concatenating their features. So we have to randomly shuffle the documents and take the adjacent _GROUP_SIZE documents in a rolling window manner to reduce the computational complexity. This is not a problem in training since each list will be shuffled differently and the model can get improved overall. So your gain in training NDCG does mean the model becomes better. The difficulty lies in Prediction and a random shuffle leads to different output for different prediction calls. This is not desired. Though in general you can call the prediction multiple times and use the mean as the final prediction. It still has variance.

Also, you may want to compare how much gain you can get by comparing _GROUP_SIZE =1 and > 1. The tensorboard can be used for this comparison.

On the other hand, we have proposed a new approach to capture cross-document interactions (https://arxiv.org/pdf/1910.09676.pdf), which is recommended. However, we don't have the implementation in our lib now and will consider to add this later.

**Comment by azagniotov** [comment]
Hi @xuanhuiwang,

Thank you for your reply and providing clarification about the `_GROUP_SIZE`.

We performed another experiment where we trained a model eight times having `_GROUP_SIZE = 1`. After each training, `NDCG@20` was calculated using ranker predictions and compared to the final `NDCG@20` observed from the validation printout during each training time.

After each training time, a model was called using the ranker object using the following code:
```
predictions = ranker.predict(input_fn=lambda: predict_input_fn(_VALIDATION_DATA_PATH)) 
```
where `_VALIDATION_DATA_PATH` contains a path to a TF record containing `600` ELWCs. The same TF record under the aforementioned path was used as an input to `eval_input_fn`.

We used the `predictions` generator and created predictions until the generator was exhausted. Each set of predictions was sorted in descending order and the same order was applied to the labels, which then were used to calculate the `NDCG@20` using the aforementioned common formula for DCG from my earlier post. We used the mean of all calculated `NDCG@20` as the final `NDCG@20` score (i.e.: `Ranker/ndcg@20` in the following summary).

The following is the `NDCG@20` comparison summary:
```
global_step = 15000: metric/ndcg@20 = 0.70970961, Ranker/ndcg@20 = 0.70970961
global_step = 15000: metric/ndcg@20 = 0.68561172, Ranker/ndcg@20 = 0.68561172
global_step = 15000: metric/ndcg@20 = 0.70615757, Ranker/ndcg@20 = 0.70615757
global_step = 15000: metric/ndcg@20 = 0.70856166, Ranker/ndcg@20 = 0.70359489
global_step = 15000: metric/ndcg@20 = 0.71078574, Ranker/ndcg@20 = 0.51706687
global_step = 15000: metric/ndcg@20 = 0.71255479, Ranker/ndcg@20 = 0.71223910
global_step = 15000: metric/ndcg@20 = 0.71907423, Ranker/ndcg@20 = 0.70144129
global_step = 15000: metric/ndcg@20 = 0.70560205, Ranker/ndcg@20 = 0.70560211
```
What we observed, that not every time the `NDCG@20` mean score computed from labels sorted by the ranker predictions matches the `NDCG@20` in the final validation printout for the same aforementioned TF record.  There is still variance present, even when the `_GROUP_SIZE = 1` is used.

Please advise

Also, regarding the cross-document interactions, I have found the following example that demonstrates how to enable a document interaction layer during training: https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_tfrecord.py#L193 . You mentioned that the implementation is currently not in the lib , with this in mind, is it a good example to follow for enabling cross-document interactions capture during training?

Thank you

**Comment by xuanhuiwang** [comment]
The only thing that I can see now is that there may be ties in your rankers prediction. Also do you have exactly 20 documents in each of your 600 ELWCs? If not, did you handle those dummy documents in your evaluation?

**Comment by azagniotov** [comment]
Hi @xuanhuiwang ,

Yes, each ELWC contains exactly 20 documents and there are no queries without any relevant documents.

**Comment by azagniotov** [comment]
Hi @xuanhuiwang ,

Thank you for your time you taken to answer my questions on the current issue. I found out that the root cause was not the ties in this case, but the mishandling of the dummy predictions in the evaluation, i.e.: your question helped to realize that.

I will close the current ticket.
